### DISCUSSION TITLE
Analog offset

### DISCUSSION THREAD

**Discussion by Wapata** [shanteacontrols/OpenDeck#144] [Ideas]
Hi !
There is a "lower CC limit" and a "upper CC limit" for the output part of the process, but there is no "lower/mini input value" and "upper/maxi input value" for the input in the  analog par of OpenDeck Configurator.
Is it possible to add theses offsets ?

I see multiple usages of this.
At first : my case. I want to use a pressure sensor and with "zero blow" it's around 0.45 (+/-0.30) volts (offset as seen on the datasheet) so i can't make my dreaming midi-wind-controller.
Secondly : mechanical limit. If you make a midi analog pedal DIY, you probably end with a pot that doesn't reach its limits. You want your pedal to stop before breaking the pot. It work also with pedal and light sensor as in Berhinger stuff.
Finally : with the dust and the age, you may find that your pot is not precise anymore at its limits.

... well it should be useful for every 3v sensor in fact... everything that is not a pot !

Thanks for making OpenDeck super-easy to use !

**Comment by paradajz** [comment]
This is currently not supported, but sounds good. I'll implement it soon.

**Reply by Wapata** [reply]
You... you make me happy !
Thank you again !!

**Reply by paradajz** [reply]
I assume you want to test this out on the Discovery board, right? Just so that I know which firmware to attach here.

**Reply by Wapata** [reply]
Yes I do !

**Comment by paradajz** [comment]
Here's the test binary:
[discovery.zip](https://github.com/shanteacontrols/OpenDeck/files/8055517/discovery.zip)

If the new options aren't visible in the UI, make sure to refresh it.

The new options are Lower and Upper ADC offset (percentage). Here's the copy/paste from the UI help text:

```
label: "Lower ADC offset",
    helpText: `Specifies lower offset percentage which is used to calculate minimum ADC value upon which MIDI
    values will be based. Useful for inputs which cannot reach minimum ADC value. If for example, the board has
    nominal ADC range 0-4095, setting this value to 10 will calculate MIDI values based on 409-4095 range (assuming
    the upper offset is 0.), that is, lower 10% of ADC range will be cut off.`,
```

```
label: "Upper ADC offset",
    helpText: `Specifies upper offset percentage which is used to calculate maximum ADC value upon which MIDI
    values will be based. Useful for inputs which cannot reach maximum ADC value. If for example, the board has
    nominal ADC range 0-4095, setting this value to 10 will calculate MIDI values based on 0-3685 range (assuming
    the lower offset is 0.), that is, upper 10% of ADC range will be cut off.`,
```

Let me know if this is what you want.

**Reply by Wapata** [reply]
I've put 75 (%) of lowest ADC offset but it doesn't work as expected.
The control change goes form 95 to 127

If 75 % it goes from 95 to 127
If 25 % it goes from 49 to 127
If 0% it goes from 49 to 127 too

It look like the formula is not applied in the right place ?

**Reply by paradajz** [reply]
Right, missed something. Fixed version attached.

[discovery.zip](https://github.com/shanteacontrols/OpenDeck/files/8055546/discovery.zip)

**Reply by Wapata** [reply]
IT WORK !
Mouahahaha tank you so much.
I will try to find out how to make a quick video of the result. :-)

The last thing missing is giving a personal name for the Midi Device (as wind controller and foot controller will have the same name it will be confusing)

**Reply by paradajz** [reply]
This can be done if you are compiling your own firmware (custom board variant). If you just want discovery board without any modifications, you can copy `discovery.yml` file under for example `my_wind_controller.yml` name, compile it and then flash that firmware. `my_wind_controller.yml` will then be its USB name.

Instructions are here:
https://github.com/shanteacontrols/OpenDeck/wiki/Building-the-OpenDeck-firmware
https://github.com/shanteacontrols/OpenDeck/wiki/Creating-custom-board-variant#building-custom-configuration

**Reply by Wapata** [reply]
Perfect and simple as rename a file... I think I will be able to do that. ;-)
If it work like this with all the board you've worked on to make them compatible you job is even more amazing that I've tough 
From "prototype" to "finished device" it's only a matter of counting the IO and find the smaller board as possible.

**Comment by paradajz** [comment]
Now a part of new release:
https://github.com/shanteacontrols/OpenDeck/releases/tag/v6.4.0

Documentation:
https://github.com/shanteacontrols/OpenDeck/wiki/Configurable-features#lower-adc-offset-section
https://github.com/shanteacontrols/OpenDeck/wiki/Configurable-features#upper-adc-offset-section
